client/notices.go

```diff
@@ -0,0 +1,200 @@
+	Keys []string
+
+	// After, if set, includes only notices that were last repeated after this time.
+	After time.Time
+}
+
+// Notice is a notification whose identity is the combination of Type and Key
+// that has occurred Occurrences number of times.
+type Notice struct {
+	ID            string            `json:"id"`
+	Type          NoticeType        `json:"type"`
+	Key           string            `json:"key"`
+	FirstOccurred time.Time         `json:"first-occurred"`
+	LastOccurred  time.Time         `json:"last-occurred"`
+	LastRepeated  time.Time         `json:"last-repeated"`
+	Occurrences   int               `json:"occurrences"`
+	LastData      map[string]string `json:"last-data,omitempty"`
+	RepeatAfter   time.Duration     `json:"repeat-after,omitempty"`
+	ExpireAfter   time.Duration     `json:"expire-after,omitempty"`
+}
+
+type NoticeType string
+
+const (
+	// Recorded whenever a change is updated: when the change is first spawned
+	// or its status is updated.
+	ChangeUpdateNotice NoticeType = "change-update"
+
+	// A custom notice reported via the Pebble client API or "pebble notify".
+	// The key and data fields are provided by the user. The key must be in
+	// the format "mydomain.io/mykey" to ensure well-namespaced notice keys.
+	CustomNotice NoticeType = "custom"
+
+	// Warnings are a subset of notices where the key is a human-readable
+	// warning message.
+	WarningNotice NoticeType = "warning"
+)
+
+type jsonNotice struct {
+	Notice
+	RepeatAfter string `json:"repeat-after,omitempty"`
+	ExpireAfter string `json:"expire-after,omitempty"`
+}
+
+// Notice fetches a single notice by ID.
+func (client *Client) Notice(id string) (*Notice, error) {
+	var jn *jsonNotice
+	_, err := client.doSync("GET", "/v1/notices/"+id, nil, nil, nil, &jn)
+	if err != nil {
+		return nil, err
+	}
+	return jsonNoticeToNotice(jn), nil
+}
+
+// Notices returns a list of notices that match the filters given in opts,
+// ordered by the last-repeated time.
+func (client *Client) Notices(opts *NoticesOptions) ([]*Notice, error) {
+	query := makeNoticesQuery(opts)
+	var jns []*jsonNotice
+	_, err := client.doSync("GET", "/v1/notices", query, nil, nil, &jns)
+	return jsonNoticesToNotices(jns), err
+}
+
+// WaitNotices returns a list of notices that match the filters given in opts,
+// telling the server to wait up to the given timeout. They are ordered by the
+// last-repeated time.
+//
+// If the timeout elapses before any matching notices arrive, it's not
+// considered an error: WaitNotices returns a nil slice and a nil error.
+func (client *Client) WaitNotices(ctx context.Context, serverTimeout time.Duration, opts *NoticesOptions) ([]*Notice, error) {
+	query := makeNoticesQuery(opts)
+	query.Set("timeout", serverTimeout.String())
+
+	// We need to use client.raw() here to pass the context through (and we
+	// don't want retries).
+	res, err := client.raw(ctx, "GET", "/v1/notices", query, nil, nil)
```

Comment:
Related to earlier comment, wE should be able to use the new API here instead of splitting doSync in two.